ext/uri/php_uriparser.c

```diff
@@ -0,0 +1,525 @@
+		} else {
+			ZVAL_NULL(retval);
+		}
+	} else {
+		ZVAL_NULL(retval);
+	}
+
+	return SUCCESS;
+}
+
+static zend_result uriparser_read_password(const uri_internal_t *internal_uri, uri_component_read_mode_t read_mode, zval *retval)
+{
+	uriparser_uris_t *uriparser_uris = (uriparser_uris_t *) internal_uri->uri;
+	UriUriA *uriparser_uri;
+	URIPARSER_READ_URI(uriparser_uri, uriparser_uris, read_mode);
+
+	if (uriparser_uri->userInfo.first != NULL && uriparser_uri->userInfo.afterLast != NULL) {
+		char *c = strchr(uriparser_uri->userInfo.first, ':');
+		if (c != NULL && uriparser_uri->userInfo.afterLast - c - 1 > 0) {
+			ZVAL_STRINGL(retval, c + 1, uriparser_uri->userInfo.afterLast - c - 1);
+		} else {
+			ZVAL_NULL(retval);
+		}
+	} else {
+		ZVAL_NULL(retval);
+	}
+
+	return SUCCESS;
+}
+
+static zend_result uriparser_read_host(const uri_internal_t *internal_uri, uri_component_read_mode_t read_mode, zval *retval)
+{
+	uriparser_uris_t *uriparser_uris = (uriparser_uris_t *) internal_uri->uri;
+	UriUriA *uriparser_uri;
+	URIPARSER_READ_URI(uriparser_uri, uriparser_uris, read_mode);
+
+	if (uriparser_uri->hostText.first != NULL && uriparser_uri->hostText.afterLast != NULL && uriparser_uri->hostText.afterLast - uriparser_uri->hostText.first > 0) {
+		if (uriparser_uri->hostData.ip6 != NULL) {
+			smart_str host_str = {0};
+
+			smart_str_appendc(&host_str, '[');
+			smart_str_appendl(&host_str, uriparser_uri->hostText.first, uriparser_uri->hostText.afterLast - uriparser_uri->hostText.first);
+			smart_str_appendc(&host_str, ']');
+
+			ZVAL_STR(retval, smart_str_extract(&host_str));
+		} else {
+			ZVAL_STRINGL(retval, uriparser_uri->hostText.first, uriparser_uri->hostText.afterLast - uriparser_uri->hostText.first);
+		}
+	} else {
+		ZVAL_NULL(retval);
+	}
+
+	return SUCCESS;
+}
+
+static zend_result uriparser_read_port(const uri_internal_t *internal_uri, uri_component_read_mode_t read_mode, zval *retval)
+{
+	uriparser_uris_t *uriparser_uris = (uriparser_uris_t *) internal_uri->uri;
+	UriUriA *uriparser_uri = uriparser_uris->uri;
+
+	if (uriparser_uri->portText.first != NULL && uriparser_uri->portText.afterLast != NULL) {
+		ZVAL_LONG(retval, strtol(uriparser_uri->portText.first, NULL, 10));
+	} else {
+		ZVAL_NULL(retval);
+	}
+
+	return SUCCESS;
+}
+
+static zend_result uriparser_read_path(const uri_internal_t *internal_uri, uri_component_read_mode_t read_mode, zval *retval)
+{
+	uriparser_uris_t *uriparser_uris = (uriparser_uris_t *) internal_uri->uri;
+	UriUriA *uriparser_uri;
+	URIPARSER_READ_URI(uriparser_uri, uriparser_uris, read_mode);
+
+	if (uriparser_uri->pathHead != NULL) {
+		const UriPathSegmentA *p;
+		smart_str str = {0};
+
+		if (uriparser_uri->absolutePath || uriIsHostSetA(uriparser_uri)) {
+			smart_str_appends(&str, "/");
+		}
+
+		smart_str_appendl(&str, uriparser_uri->pathHead->text.first, (int) ((uriparser_uri->pathHead->text).afterLast - (uriparser_uri->pathHead->text).first));
```

Comment:
I don't like the int cast, it should probably either be a size_t or ptrdiff_t cast

Reply:
Yes, makes sense. I'll choose size_t because the result is supposed to be unsigned.